### ISSUE TITLE
Add GraphQLContext derive

### ISSUE THREAD

**Issue by LegNeato** [graphql-rust/juniper#329]
**Is your feature request related to a problem? Please describe.**
Contexts have some boilerplate, you have to implement `juniper::Context` which is usually an empty marker trait. The error message when you don't isn't obvious, see  https://github.com/graphql-rust/juniper/issues/327. 

**Describe the solution you'd like**
It might be good to have a `#[derive(GraphQLContext)` that does the boilerplate for you. It won't help the error message, but perhaps when people are copying and pasting or reading docs they will be less likely to miss it.

**Comment by theduke** [comment]
Actually, I think we could just deprecate the `Context` trait completely and remove the trait bound.

There is no real reason for it to be there.

**Comment by theduke** [comment]
I guess the only reason for it's existence would be that you can't use a type as context in the macros that isn't intended to be used as a context. 

But that would produce compile time errors anyway once you actually construct the schema.

**Comment by LegNeato** [comment]
Do we want to reserve the trait for future extensibility? If we remove the requirement and later need it for implementation specifics, it would be a breaking change.

**Comment by davidpdrsn** [comment]
@LegNeato what sort of future extensibility are you thinking of? I imagine adding methods to the trait would also be breaking change.

I'm also in favour of deprecating it.

**Comment by theduke** [comment]
I don't see a reason to ever add methods on context.
More state keeping or functionality would be added to the `Executor`.

There is a consideration with futures support though: we will probably need a `Send + Sync + Clone` bound on context!
With that in mind, we may want to add those constraints onto the `Context` trait instead of litterling them around the code everywhere.

**Comment by davidpdrsn** [comment]
I'm not that familiar with futures, but why `Clone`?

The app I'm working on has a context like this:

```rust
pub struct Context {
    db: DbCon,
    // more things...
}
```

Where `DbCon` is

```rust
pub struct DbCon(pub PooledConnection<ConnectionManager<PgConnection>>);
```

`PooledConnection` comes from [`r2d2`](https://docs.rs/r2d2/0.8.4/r2d2/struct.PooledConnection.html) and is _not_ `Clone`. So I can't easily make this context `Clone`.

I could instead store the connection pool itself in the context. The pool is `Clone` because it just [wraps and `Arc<SharedPool>`](https://docs.rs/r2d2/0.8.4/src/r2d2/lib.rs.html#316-318). That feels like it would create a lot of churn on the pool though. If each db query would have to first get a new connection.

I also see that `PgConnection` from Diesel is also [not `Sync`](https://docs.rs/diesel/1.4.2/diesel/pg/struct.PgConnection.html#impl-Sync) 🤔

**Comment by theduke** [comment]
> but why `Clone`?

That's a necessary result from concurrency/parallelism. 
We'll need to clone the Context often.

You could still use a single DB connection per query by wrapping it in `Arc<Mutex<Connection>>`.

Note that the diesel connections are not `Sync`, but are `Send`.

**Comment by davidpdrsn** [comment]
Makes sense. Seems to still be [early days for async in Diesel](https://github.com/diesel-rs/diesel/issues/399).

**Comment by petoknm** [comment]
I think we could use a trait alias for this in the future for async. Having an empty trait doesn't make much sense to me. I will try to make a PR for this.

**Comment by theduke** [comment]
@petoknm great that you want to help, I would hold off on this for now though until I publish my async/await branch and we have a clearer picture of what will be needed.

**Comment by petoknm** [comment]
Cool, I didn't know that work on async support has already started.
